### PR TITLE
Refactor the Parser and Compiler classes

### PR DESCRIPTION
**Pull request by codingpotato**
Refactor some code to separate `Compiler` and `ExprCompiler`, which can make future development easier.

There isn't any functional change in the PR.

### CODE REVIEW

File: csharp/src/SeedLang/Ast/AstDumper.cs [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+          return "%";
+        default:
+          throw new NotImplementedException($"Unsupported binary operator: {op}.");
+      }
+    }
+
+    // Returns the internal string representation of boolean operators.
+    internal static string Symbol(this BooleanOperator op) {
+      switch (op) {
+        case BooleanOperator.And:
```
Comment (wixette) [inline csharp/src/SeedLang/Ast/AstDumper.cs:46]: "And", "Or" and "Not: both Python and Lua use lowercase form for those operators - "and", "or", "not". Shall we follow this convention for internal representations of them?

File: csharp/src/SeedLang/Engine.cs [hunk 2]
```diff
@@ -34,8 +34,8 @@ public class Engine {
     private readonly RunMode _runMode;
     private readonly VM _vm = new VM();
 
-    // The AST tree of the source code.
-    private AstNode _astNode;
+    // The AST tree of the program.
+    private Statement _program;
```
Comment (wixette) [inline csharp/src/SeedLang/Engine.cs:38]: Why don't name this variable "_astTree" or "_programAsAst" ?

Comment (codingpotato) [reply csharp/src/SeedLang/Engine.cs]: Fixed.

File: csharp/src/SeedLang/Ast/AstDumper.cs (hunk 1 above)
Comment (codingpotato) [reply csharp/src/SeedLang/Ast/AstDumper.cs:46]: Fixed.

File: csharp/src/SeedLang/Ast/AstDumper.cs [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+    // Returns the internal string representation of unary operators.
+    internal static string Symbol(this UnaryOperator op) {
+      switch (op) {
+        case UnaryOperator.Positive:
+          return "+";
+        case UnaryOperator.Negative:
+          return "-";
+        case UnaryOperator.Not:
+          return "Not";
```
Comment (wixette) [inline csharp/src/SeedLang/Ast/AstDumper.cs:85]:
```suggestion
          return "not";
```

**Review by wixette — APPROVED** [review]
LGTM

**Pull request merged**